Add app host configuration page (#2141)
dotnet/docs-aspire: docs/app-host/configuration.md

---
title: .NET Aspire app host configuration

> description: .NET Aspire app host configuration options

Reviewer: ```suggestion
description: Learn about the .NET Aspire app host configuration options.
```

ms.date: 11/21/2024
ms.topic: reference
---

# App host configuration

The app host project configures and starts your distributed application (<xref:Aspire.Hosting.DistributedApplication>). When a `DistributedApplication` runs it reads configuration from the app host. Configuration is loaded from environment variables that are set on the app host and <xref:Aspire.Hosting.DistributedApplicationOptions>.

Configuration includes:

- Settings for hosting the resource service, such as the address and authentication options.

> - Settings used to start the [Aspire dashboard](../fundamentals/dashboard/overview.md), such the dashboard's frontend and OpenTelemetry Protocol (OTLP) addresses.

Reviewer: ```suggestion
- Settings used to start the [.NET Aspire dashboard](../fundamentals/dashboard/overview.md), such the dashboard's frontend and OpenTelemetry Protocol (OTLP) addresses.
```

- Internal settings that Aspire uses to run the app host. These are set internally but can be accessed by integrations that extend Aspire.

App host configuration is provided by the app host launch profile. The app host has a launch settings file call `launchSettings.json` which has a list of launch profiles. Each launch profile is a collection of related options which defines how you would like `dotnet` to start your application.

```json
{
  "$schema": "https://json.schemastore.org/launchsettings.json",
  "profiles": {
    "https": {
      "commandName": "Project",
      "dotnetRunMessages": true,
      "launchBrowser": true,
      "applicationUrl": "https://localhost:17134;http://localhost:15170",
      "environmentVariables": {
        "ASPNETCORE_ENVIRONMENT": "Development",
        "DOTNET_ENVIRONMENT": "Development",
        "DOTNET_DASHBOARD_OTLP_ENDPOINT_URL": "https://localhost:21030",
        "DOTNET_RESOURCE_SERVICE_ENDPOINT_URL": "https://localhost:22057"
      }
    }
  }
}
```

> The proceeding launch settings file:

Reviewer: ```suggestion
The preceding launch settings file:
```

> - Has one launch profile called `https`.

Reviewer: ```suggestion
- Has one launch profile named `https`.
```

- Configures an .NET Aspire app host project:
  - The `applicationUrl` property configures the dashboard launch address (`ASPNETCORE_URLS`).

>   - Env vars such as `DOTNET_DASHBOARD_OTLP_ENDPOINT_URL` and `DOTNET_RESOURCE_SERVICE_ENDPOINT_URL` are set on the app host.

Reviewer: ```suggestion
  - Environment variables such as `DOTNET_DASHBOARD_OTLP_ENDPOINT_URL` and `DOTNET_RESOURCE_SERVICE_ENDPOINT_URL` are set on the app host.
```

For more information, see [.NET Aspire and launch profiles](../fundamentals/launch-profiles.md).

> [!NOTE]
> Configuration described on this page is for .NET Aspire app host project. To configure the standalone dashboard, see [dashboard configuration](../fundamentals/dashboard/configuration.md).

## Common configuration

> | Option | Default Value | Description |

Reviewer: ```suggestion
| Option | Default value | Description |
```

|--|--|--|
| `ASPIRE_ALLOW_UNSECURED_TRANSPORT` | `false` | Allows communication with the app host without https. `ASPNETCORE_URLS` (dashboard address) and `DOTNET_RESOURCE_SERVICE_ENDPOINT_URL` (app host resource service address) must be secured with HTTPS unless true. |

## Resource service

> 

Reviewer: It would be nice to have an introductory sentence here, like you've done with some of the other sections. What is the resource service and why would I want to configure it?

> | Option | Default Value | Description |

Reviewer: ```suggestion
| Option | Default value | Description |
```

|--|--|--|

> | `DOTNET_RESOURCE_SERVICE_ENDPOINT_URL` | Automatically generated 128-bit entropy token. | Configures the address of the resource service hosted by the app host. Automatically generated with `launchSettings.json` to have a random port on localhost. For example, `https://localhost:17037`. |

Reviewer: ```suggestion
| `DOTNET_RESOURCE_SERVICE_ENDPOINT_URL` | Automatically generated 128-bit entropy token. | Configures the address of the resource service hosted by the app host. Automatically generated with _launchSettings.json_ to have a random port on localhost. For example, `https://localhost:17037`. |
```

| `DOTNET_DASHBOARD_RESOURCESERVICE_APIKEY` | `null` | The API key used to authenticate requests made to the app host's resource service. The API key is required if the app host is in run mode, the dashboard isn't disabled, and the dashboard isn't configured to allow anonymous access with `DOTNET_DASHBOARD_UNSECURED_ALLOW_ANONYMOUS`. |

## Dashboard

By default, the dashboard is automatically started by the app host. The dashboard supports [its own set of configuration](../fundamentals/dashboard/configuration.md), and some settings can be configured from the app host.

> | Option | Default Value | Description |

Reviewer: ```suggestion
| Option | Default value | Description |
```

|--|--|--|

> | `ASPNETCORE_URLS` | `null` | Dashboard address. Must be https unless `ASPIRE_ALLOW_UNSECURED_TRANSPORT` or `DistributedApplicationOptions.AllowUnsecuredTransport` is true. Automatically generated with `launchSettings.json` to have a random port on localhost. The value in launch settings is set on the `applicationUrls` property. |

Reviewer: Alternatively, if not `https` then HTTPS. Also, all references to files should be italicized.

```suggestion
| `ASPNETCORE_URLS` | `null` | Dashboard address. Must be `https` unless `ASPIRE_ALLOW_UNSECURED_TRANSPORT` or `DistributedApplicationOptions.AllowUnsecuredTransport` is true. Automatically generated with _launchSettings.json_ to have a random port on localhost. The value in launch settings is set on the `applicationUrls` property. |
```

> | `ASPNETCORE_ENVIRONMENT` | `Production` | Configures the environment the dashboard runs as. For more information, see [Use multiple environments in ASP.NET Core](https://learn.microsoft.com/aspnet/core/fundamentals/environments). |

Reviewer: ```suggestion
| `ASPNETCORE_ENVIRONMENT` | `Production` | Configures the environment the dashboard runs as. For more information, see [Use multiple environments in ASP.NET Core](/aspnet/core/fundamentals/environments). |
```

> | `DOTNET_DASHBOARD_OTLP_ENDPOINT_URL` | `http://localhost:18889` if no gRPC endpoint is configured. | Configures the dashboard OTLP gRPC address. Used by the dashboard to receive telemetry over OTLP. Set on resources as the `OTEL_EXPORTER_OTLP_ENDPOINT` env var. The `OTEL_EXPORTER_OTLP_PROTOCOL` env var is `grpc`.  Automatically generated with `launchSettings.json` to have a random port on localhost. |

Reviewer: ```suggestion
| `DOTNET_DASHBOARD_OTLP_ENDPOINT_URL` | `http://localhost:18889` if no gRPC endpoint is configured. | Configures the dashboard OTLP gRPC address. Used by the dashboard to receive telemetry over OTLP. Set on resources as the `OTEL_EXPORTER_OTLP_ENDPOINT` env var. The `OTEL_EXPORTER_OTLP_PROTOCOL` env var is `grpc`.  Automatically generated with _launchSettings.json_ to have a random port on localhost. |
```

| `DOTNET_DASHBOARD_OTLP_HTTP_ENDPOINT_URL` | `null` | Configures the dashboard OTLP HTTP address. Used by the dashboard to receive telemetry over OTLP. If only `DOTNET_DASHBOARD_OTLP_HTTP_ENDPOINT_URL` is configured then it is set on resources as the `OTEL_EXPORTER_OTLP_ENDPOINT` env var. The `OTEL_EXPORTER_OTLP_PROTOCOL` env var is `http/protobuf`. |
| `DOTNET_DASHBOARD_FRONTEND_BROWSERTOKEN` | Automatically generated 128-bit entropy token. | Configures the frontend browser token. This is the value that must be entered to access the dashboard when the auth mode is BrowserToken. If no browser token is specified then a new token is generated each time the app host is launched. |

## Internal

Internal settings are used by the app host and integrations. Internal settings aren't designed to be configured directly.

> | Option | Default Value | Description |

Reviewer: ```suggestion
| Option | Default value | Description |
```

|--|--|--|

> | `AppHost:Directory` | The content root if there's no project. | Directory of the project where the app host is located. |

Reviewer: Is this exposed through the distributed app builder as the [`IDistributedApplicationBuider.AppHostDirectory`](https://learn.microsoft.com/en-us/dotnet/api/aspire.hosting.idistributedapplicationbuilder.apphostdirectory?view=dotnet-aspire-9.0#aspire-hosting-idistributedapplicationbuilder-apphostdirectory)? Might be good to add an xref to that.
```suggestion
| `AppHost:Directory` | The content root if there's no project. | Directory of the project where the app host is located. Accessible from the <xref:Aspire.Hosting.IDistributedApplicationBuilder.AppHostDirectory?displayProperty=nameWithType>. |
```
